### PR TITLE
[GPU-plugin] added GPU objective function

### PR DESCRIPTION
**Pull request by teju85**
Objective functions can now also be run on GPUs.
Currently, reg:linear, reg:logistic, binary:logistic and binary:logitraw functions have been ported.
To select their GPU counterparts, prefix 'gpu:' in their names. (Refer to parameter.md for more info)

### CODE REVIEW

**Comment by khotilov** [comment]
To keep things organized, could these be renamed:
loss.h -> regression_loss.h
objective_gpu.cu -> regression_obj_gpu.cu

File: src/objective/objective.cc [hunk 1]
```diff
@@ -28,6 +28,7 @@ namespace xgboost {
 namespace obj {
 // List of files that will be force linked in static links.
 DMLC_REGISTRY_LINK_TAG(regression_obj);
+DMLC_REGISTRY_LINK_TAG(objective_gpu);
```
Comment (RAMitchell) [inline src/objective/objective.cc]: Use the macro #ifdef XGBOOST_USE_CUDA around this registry link tag

File: src/objective/objective_gpu.cu [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= n)
+    return;
+  float p = Loss::PredTransform(preds[i]);
+  float w = weights == nullptr ? 1.0f : weights[i];
+  float label = labels[i];
+  if (label == 1.0f)
+    w *= scale_pos_weight;
+  if (!Loss::CheckLabel(label))
+    *label_correct = false;
```
Comment (RAMitchell) [inline src/objective/objective_gpu.cu]: Is it safe to set label_correct like this?

File: src/objective/objective_gpu.cu [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  protected:
+  
+  float *preds_d_, *labels_d_, *weights_d_;
+  bool *label_correct_d_;
+  bst_gpair *out_gpair_d_;
+  int max_n_;
+  std::vector<cudaStream_t> streams_;
+  bool copied_;
+
+  // allocate device data for n elements
```
Comment (RAMitchell) [inline src/objective/objective_gpu.cu]: Consider using the dvec and bulk allocator classes in common/device_helpers.cuh to manage device memory.

**Comment by RAMitchell** [comment]
Can you please provide some benchmarks showing relative speed improvement. I was not aware gradient calculation was a significant bottleneck. Also please tidy up the commit history and deal with CI failures.

**Comment by RAMitchell** [comment]
Closing as no updates.

**Comment by teju85** [comment]
My bad. I got distracted with other issues and couldn't get to this in time.
We already have these feedbacks incorporated in an internal fork.
I'll file another PR with these soon.
